pyephember/pyephember.py

```diff
@@ -45,39 +118,178 @@ def zone_name(zone):
+    from the PointIndex enum: 'ADVANCE_ACTIVE', 'CURRENT_TEMP', etc
+    """
+    # pylint: disable=unsubscriptable-object
+    if hasattr(PointIndex, index):
+        index = PointIndex[index].value
+
+    for datum in zone['pointDataList']:
+        if datum['pointIndex'] == index:
+            return int(datum['value'])
+
+    return None
 
 
 def zone_mode(zone):
     """
     Get mode for this zone
     """
-    return ZoneMode(zone['mode'])
+    return ZoneMode(zone_pointdata_value(zone, 'MODE'))
+
+
+class EphMessenger:
+    """
+    MQTT interface to the EphEmber API
+    """
+
+    def _zone_command_b64(self, zone, cmd, stop_mqtt=True, timeout=1):
+        """
+        Send a base64-encoded MQTT command to a zone
+        Returns true if the command was published within the timeout
+        """
+        product_id = self.parent.get_home_details()['homes']['productId']
+        uid = self.parent.get_home_details()['homes']['uid']
+
+        msg = json.dumps(
+            {
+                "common": {
+                    "serial": 7870,
+                    "productId": product_id,
+                    "uid": uid,
+                    "timestamp": str(int(1000*time.time()))
+                },
+                "data": {
+                    "mac": zone['mac'],
+                    "pointData": cmd
+                }
+            }
+        )
+
+        started_locally = False
+        if not self.client or not self.client.is_connected():
+            started_locally = True
+            self.start()
+
+        pub = self.client.publish(
+            "/".join([product_id, uid, "download/pointdata"]), msg, 0
+        )
+        pub.wait_for_publish(timeout=timeout)
+
+        if started_locally and stop_mqtt:
+            self.stop()
+
+        return pub.is_published()
+
+    # Public interface
+
+    def start(self, callbacks=None, loop_start=False):
+        """
+        Start MQTT client
+        """
+        credentials = self.parent.messenging_credentials()
+        self.client_id = '{}_{}'.format(
+            credentials['user_id'], str(int(1000*time.time()))
+        )
+        token = credentials['token']
+
+        mclient = mqtt.Client(self.client_id)
+        mclient.tls_set()
+        self.client = mclient
+
+        user_name = "app/{}".format(token)
+        mclient.username_pw_set(user_name, token)
+
+        if callbacks is not None:
+            for key in callbacks.keys():
+                setattr(mclient, key, callbacks[key])
+
+        mclient.connect(self.api_url, self.api_port)
+
+        if loop_start:
+            mclient.loop_start()
+
+        return mclient
+
+    def stop(self):
+        """
+        Disconnect MQTT client if connected
+        """
+        if not self.client:
+            return False
+        if self.client.is_connected():
+            self.client.disconnect()
+        return True
+
+    def zone_command(self, zone, ints_cmd, stop_mqtt=True, timeout=1):
```

Comment:
It would be nice to have a bit of a better abstraction over the ints_cmds array. Could something like small class or namedtuple be used here to make it a bit nicer?

Reply:
Agreed, I'll try to smooth that over when I get a moment.